### PR TITLE
[MISC] Check for peer relation when getting the primary endpoint

### PR DESCRIPTION
**Pull request by dragomirp**
Try to speed up and stabilise the tests. Based of the shared buffers PR.

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -299,6 +300,9 @@ def postgresql(self) -> PostgreSQL:
     @property
     def primary_endpoint(self) -> Optional[str]:
         """Returns the endpoint of the primary instance or None when no primary available."""
+        if not self._peers:
+            logger.debug("primary endpoint early exit: Peer relation not joined yet.")
+            return None
```
Comment (dragomirp) [inline src/charm.py:305]: If the peer relation is not yet joined, don't try to find the primary.

File: src/charm.py [hunk 2]
```diff
@@ -1036,6 +1053,10 @@ def _start_primary(self, event: StartEvent) -> None:
             logger.exception(e)
             self.unit.status = BlockedStatus("Failed to create postgres user")
             return
+        except PostgreSQLListUsersError:
+            logger.warning("Deferriing on_start: Unable to list users")
+            event.defer()
+            return
```
Comment (dragomirp) [inline src/charm.py:1059]: Skipping the wait in primary_endpoint could lead o exceptions here.

File: src/charm.py [hunk 3]
```diff
@@ -1424,6 +1456,10 @@ def update_config(self, is_creating_backup: bool = False) -> bool:
             logger.debug("Early exit update_config: Patroni not started yet")
             return False
 
+        # Try to connect
+        if not self._can_connect_to_postgresql:
+            logger.debug("Early exit update_config: Cannot connect to Postgresql")
+            return False
```
Comment (dragomirp) [inline src/charm.py:1462]: If the node's db is not yet up, we can't really check those values. Next update status should run this again.

File: src/charm.py [hunk 4]
```diff
@@ -308,7 +312,20 @@ def primary_endpoint(self) -> Optional[str]:
...
+                        # TODO figure out why peer data is not available
+                        if (
+                            primary_endpoint
+                            and len(self._units_ips) == 1
+                            and len(self._peers.units) > 1
+                        ):
+                            logger.warning(
+                                "Possibly incoplete peer data: Will not map primary IP to unit IP"
+                            )
+                            return primary_endpoint
```
Comment (dragomirp) [inline src/charm.py:324]: On juju 3 we don't seem to be getting the peer data used to get the IP addr when a new unit is joining (e.g. when scaling in the `test_db` suite).

File: src/charm.py [hunk 5]
```diff
@@ -308,7 +312,20 @@ def primary_endpoint(self) -> Optional[str]:
...
+                        logger.debug(
+                            "primary endpoint early exit: Primary IP not in cached peer list."
+                        )
+                        primary_endpoint = None
```
Comment (dragomirp) [inline src/charm.py:328]: `_units_ips` is a property and will not update on retry. The Patroni response might change while retrying, but IMHO it's better to leave that to the topology observer script.

File: tests/integration/test_db_admin.py [hunk 6]
```diff
@@ -34,7 +34,6 @@
 
 
 @pytest.mark.group(1)
-@pytest.mark.skip(reason="DB Admin tests are currently broken")
```
[the remark this answers is not in the document]
Comment (dragomirp) [reply tests/integration/test_db_admin.py:37]: I reenabled it here to reduce the amount CI runs.

File: src/charm.py (hunk 3 above)
Comment (dragomirp) [reply src/charm.py]: Changed to warning.

**Review by taurus-forever — APPROVED** [review]
Discussed with Dragomir, merging this to have progress with tests stabilisation otherwise shared_buffers test cannot be merged.

**Pull request merged**